[RFC] VPC construct

t-richard opened the pull request:
This is a draft PR related to the VPC RFC.

This implementation is far from complete and is here more to open a discussion than to be merged immediatly.

TODO

- [x] Add proper tests
- [ ] ~~support more options~~
- [ ] ~~add references~~
- [ ] ~~add outputs~~
- [ ] documentation
- [x] properly handle account/regions/AZs (only support 2AZ for now so we are in an environment-agnostic setup)

mnapoli commented:
Hey @t-richard, turns out I already have a WIP of a VPC construct 😅  It looks very similar btw, and it will go hand in hand with an RDS construct.

I just want to let you know before you invest much more work into this PR.

mnapoli commented:
Following discussions in the RFC and with Frédéric: let's go with your solution (having a proper "VPC" construct). Don't bother rebasing/incorporating my branch.

In order to start with a limited scope:

- it should be production-ready: 2 AZ, 1 NAT per AZ
- caring about a "dev" mode is out of scope for now
- because of that, let's ignore making the gateway type customizable for now (NAT Gateway only)

And more generally: don't add anything we don't have a use case for. That includes variables, outputs, etc. The reason is that it's easier to add anything later, but it's extremely hard to remove anything.

Let me know if I can help in any way!

mnapoli commented on src/classes/AwsProvider.ts:
```diff
@@ -84,6 +88,17 @@ export class AwsProvider {
         this.serverless.service.setFunctionNames(this.serverless.processedInput.options);
     }
 
+    setVpcConfig(securityGroup: AwsCfInstruction, subnets: AwsCfInstruction[]): void {
```
This should probably error if there's already a VPC setup?

t-richard replied:
Yes, it should. And we could handle the "2 VPC constructs" case here too.

mnapoli commented on src/constructs/VPC.ts:
```diff
@@ -0,0 +1,76 @@
+        natGatewayType: {
+            anyOf: [{ const: "gateway" }, { const: "instance" }],
+        },
+        natGateways: { type: "number" },
```
I think we can remove those 2 settings for now.

mnapoli commented on test/unit/vpc.test.ts:
```diff
@@ -0,0 +1,25 @@
+    Outputs?: Record<string, unknown>;
+};
+
+function expectUserDynamoStatementIsAdded(cfTemplate: CfTemplate) {
```
Is "Dynamo" a leftover from another branch?

t-richard replied:
yep, definitly copy/pasted some test code and did not cleaned-it-up :sweat_smile:

t-richard commented:
> In order to start with a limited scope:
>
> it should be production-ready: 2 AZ, 1 NAT per AZ
caring about a "dev" mode is out of scope for now
because of that, let's ignore making the gateway type customizable for now (NAT Gateway only)

That makes sense, thanks for the feedback. 

I'll try to work on that this week. This is almost a working solution for the described use case but it needs a bit or rework, more tests and things like that.

mnapoli commented on src/classes/AwsProvider.ts:
```diff
@@ -84,6 +85,35 @@ export class AwsProvider {
     }
 
+    /**
+     * @internal
```
Good idea, better err on the safe side first.

mnapoli commented on src/classes/AwsProvider.ts:
```diff
@@ -84,6 +85,35 @@ export class AwsProvider {
+        }
+
+        this.serverless.service.provider.vpc = {
+            securityGroupIds: [securityGroup], // TODO : merge with existing groups ?
```
I think the TODO can be removed now? (if VPC settings are defined in serverless.yml then there will be an error)

t-richard replied:
That's more of an idea for later. 

Security groups are not related to a specific subnet/vpc so one may want to provide a custom security group that is shared accross projets/environments and we might want to merge our newly created one with those provided by the user.

mnapoli replied:
👍  ok makes sense

mnapoli commented on src/classes/AwsConstruct.ts:
```diff
@@ -29,5 +30,5 @@ export abstract class AwsConstruct extends CdkConstruct implements ConstructInte
      * CloudFormation references
      */
-    abstract references(): Record<string, Record<string, unknown>>;
+    abstract references(): Record<string, AwsCfInstruction>;
```
These changes have been merged on master, it looks like a rebase problem?

t-richard replied:
Wasn't aware of that, I applied those changes my self. I will try to rebase.

t-richard replied:
I don't see those on master. It should be here for example https://github.com/getlift/lift/blob/master/src/classes/AwsConstruct.ts#L32

Let me know if I'm missing something.

mnapoli replied:
🤦  my bad those were in the `db` branch. All good!

mnapoli commented on src/constructs/VPC.ts:
```diff
@@ -0,0 +1,52 @@
+
+        const privateSubnets = this.vpc.selectSubnets({
+            subnetType: SubnetType.PRIVATE,
+        }).subnetIds;
```
There is the `vpc.privateSubnets` property, is there any difference? (I'm just arguing for `vpc.privateSubnets` because it is simpler)

Mmh after looking up, and unless I'm mistaken, it seems to be equivalent: https://github.com/aws/aws-cdk/blob/283ed02c64f827161edba3e11c3cead3b54b7ee9/packages/%40aws-cdk/aws-ec2/lib/vpc.ts#L524-L529

t-richard replied:
They indeed seem similar, so let's go with the more concise `vpc.privateSubnets` syntax.

mnapoli commented on test/unit/vpc.test.ts:
```diff
@@ -0,0 +1,75 @@
+import {AwsCfInstruction} from "@serverless/typescript";
+
+describe("vpc", () => {
+    it("should have subnets and security groups", async () => {
```
If I understand correctly what this test creates, it could be improved to:

```suggestion
    it("should put Lambda functions in the VPC", async () => {
```

mnapoli commented on test/unit/vpc.test.ts:
```diff
@@ -0,0 +1,75 @@
+            expect(error).toBeInstanceOf(ServerlessError);
+            expect(error).toHaveProperty("code", "LIFT_ONLY_ONE_VPC");
+        }
+    });
```
Awesome testing the error cases. I would also add a `should create all required resources` kind-of test, like for the other construct: i.e. test that `serverless deploy` actually deploys a VPC with all the related resources in the CloudFormation template.

t-richard replied:
Isn't it testing the CDK more than testing Lift ? I'll see what I can come up with.

mnapoli replied:
👍  have a look at the other constructs

The idea is to test what is being deployed, no matter if it's using the default values of CDK constructs or if we customize it. Lift provides the outcome, and unlike CloudFormation CDK constructs could evolve.

t-richard replied:
Done in the latest commit

mnapoli reviewed (commented):
That looks awesome, thank you!

I've added minor comments and then for me we'll be good to merge!

t-richard commented on src/constructs/VPC.ts:
```diff
@@ -0,0 +1,52 @@
+            subnetType: SubnetType.PRIVATE,
+        }).subnetIds;
+
+        const lambdaSecurityGroup = new SecurityGroup(this, "AppSecurityGroup", {
```
I think we need to add an egress here to allow outbound trafic on http/https

mnapoli replied:
Yes, I would even say allow all outbound traffic.

t-richard replied:
Done in the latest commit

t-richard commented on src/constructs/VPC.ts:
```diff
@@ -0,0 +1,50 @@
+            maxAzs: 2,
+        });
+
+        const privateSubnets = this.vpc.privateSubnets;
```
Also maybe we would like to create isolated subnets to put services into (like RDS, ElastiCache, etc), instead of putting them in the private subnets.

mnapoli replied:
Let's put that aside for now, putting these services in the private subnets should be a good fit for most use cases.

@deleugpn mentioned that isolated subnets could be useful for security compliance, but these are advanced use cases.

t-richard commented:
Added tests to ensure resources are created and added the security group egress.

Also removed the calls to `provider.getCloudformationReference()` to be ready for https://github.com/getlift/lift/pull/71

I think the only missing thing is docs now. I'll work on that tonight and we should be good to go.

mnapoli commented:
As discussed in Slack, merged as a "shadow" construct: it is not documented. It is experimental for now, and will be iterated as we identify use cases around it (e.g. RDS, etc.).
